sdk/metric/integrator/simple/simple.go

```diff
@@ -68,97 +84,209 @@ type (
+			processStart:  now,
+			intervalStart: now,
 		},
 	}
 }
 
-func (b *Integrator) Process(accumulation export.Accumulation) error {
+// Process implements export.Integrator.
+func (b *Integrator) Process(accum export.Accumulation) error {
 	if b.startedCollection != b.finishedCollection+1 {
 		return ErrInconsistentState
 	}
-
-	desc := accumulation.Descriptor()
-	key := batchKey{
+	desc := accum.Descriptor()
+	key := stateKey{
 		descriptor: desc,
-		distinct:   accumulation.Labels().Equivalent(),
-		resource:   accumulation.Resource().Equivalent(),
-	}
-	agg := accumulation.Aggregator()
-	value, ok := b.batch.values[key]
-	if ok {
-		// Note: The call to Merge here combines only
-		// identical accumulations.  It is required even for a
-		// stateless Integrator because such identical accumulations
-		// may arise in the Meter implementation due to race
-		// conditions.
-		return value.aggregator.Merge(agg, desc)
-	}
-	// If this integrator is stateful, create a copy of the
-	// Aggregator for long-term storage.  Otherwise the
-	// Meter implementation will checkpoint the aggregator
-	// again, overwriting the long-lived state.
-	if b.stateful {
-		tmp := agg
-		// Note: the call to AggregatorFor() followed by Merge
-		// is effectively a Clone() operation.
-		b.AggregatorFor(desc, &agg)
-		if err := agg.Merge(tmp, desc); err != nil {
-			return err
+		distinct:   accum.Labels().Equivalent(),
+		resource:   accum.Resource().Equivalent(),
+	}
+	agg := accum.Aggregator()
+
+	// Check if there is an existing record.
+	value, ok := b.state.values[key]
+	if !ok {
+		stateful := b.ExportKindFor(desc, agg.Aggregation().Kind()).MemoryRequired(desc.MetricKind())
+
+		newValue := &stateValue{
+			labels:   accum.Labels(),
+			resource: accum.Resource(),
+			updated:  b.state.finishedCollection,
+			stateful: stateful,
+			current:  agg,
+		}
+		if stateful {
+			// If stateful, allocate a cumulative aggregator.
+			b.AggregatorFor(desc, &newValue.cumulative)
+
+			if desc.MetricKind().PrecomputedSum() {
+				// If we need to compute deltas, allocate another aggregator.
+				b.AggregatorFor(desc, &newValue.delta)
+			}
 		}
+		b.state.values[key] = newValue
+		return nil
 	}
-	b.batch.values[key] = batchValue{
-		aggregator: agg,
-		labels:     accumulation.Labels(),
-		resource:   accumulation.Resource(),
+
+	// Advance the update sequence number:
+	sameRound := b.state.finishedCollection == value.updated
+	value.updated = b.state.finishedCollection
+
+	// An existing record will be found when:
+	// (a) stateful aggregation is required for an exporter
+	// (b) multiple accumulators (SDKs) are being used.
+	// Another accumulator must have produced this.
+
+	if !sameRound {
+		// This is the first time through in a new round.
+		value.current = agg
+		return nil
 	}
-	return nil
+	if desc.MetricKind().Asynchronous() {
+		// The last value across multiple accumulators is taken.
+		value.current = agg
+		return nil
+	}
+	if value.delta == nil {
```

Comment:
I'm attempting to review this PR with limited knowledge of the SDK pipeline, so it may just be I'm lacking context, but I find it very hard to figure out what's happening in this function.

I think part of the difficulty for me is distinguishing between work that's done to convert between cumulative & delta values, and work that's done to aggregate for other reasons (when multiple accumulators are used?). In particular, I thought the `delta` aggregation was only needed if we had a pre-computed sum that needs to be converted back to a delta for the exporter (as per line 138), so I'm not sure why we are using it here.

Is it possible to separate this logic more clearly? I'd try to make a suggestion of how to go about that, but not sure I'm following along well enough.

Reply:
I agree, this could be clearer.

I feel like breaking apart to separate functions could help. That, or really explicit comments targeted at devs with no knowledge of the pipeline being added.

Reply:
Thanks for adding comments. It's definitely easier to understand now.

I still think some code changes could improve readability as well though. In particular, do we have to use the `delta` aggregator property for both storing the conversion from `cumulative` to `delta` AND for storing the multi-aggregator temp results? In fact, can we not just store the multi-aggregator temp result in `current` & use a flag to determine if we have performed the Clone? (which would happen when we see the second accumulator in the same collection)

Reply:
You are right, we do not have to use the `delta` aggregator in two different ways.  If we use `current` and a flag, it won't be possible to keep the state that was allocated for storing `delta` values (in both cases). The way this is coded allows referring to an Aggregator owned by the Accumulator in the common case of a single Accumulator, as well as the less-common case of a stateful Asynchronous instrument.

I will also say that in #799 I tried doing this in another way, and it was even worse. I'm sure this can be improved, and I want to assure you that you seem to understand what's happening very well!